### PR TITLE
Fine-tuning Adaptation for QVED Dataset

### LINKED ISSUE #2: Fix VideoLLaMA3 Training Failures
Adapting VideoLLaMA3 for fine-tuning on the QVED exercise form correction dataset requires fixing multiple critical issues related to dataset format, memory management, and inference pipeline compatibility.

#### 1. **Dataset Format Incompatibility** 
- **Problem**: Training script expected video paths as lists `["video.mp4"]`, dataset had strings `"video.mp4"`
- **Problem**: Missing `<video>` tag in conversation prompts caused video loading failures

#### 2. **Multi-GPU Configuration on Single GPU** 
- **Problem**: Default config used `NPROC_PER_NODE=8` causing "ChildFailedError" and Flash Attention warnings

#### 3. **Out of Memory (OOM) Errors** 
- **Problem**: Training crashed with CUDA OOM on 48GB A40 GPU

#### 4. **Missing Dependencies** 
- **Problem**: `ffprobe` not found, causing video loading failures

#### 5. **Checkpoint Not Saving** 
- **Problem**: Training completed but model wasn't saved (using `save_steps=20` with 21 total steps)

#### 6. **Inference Pipeline Incompatibility** 
- **Problem**: Old Mobile-VideoGPT inference code incompatible with VideoLLaMA3 API

#### 7. **Token Sequence Length Overflow** 
- **Problem**: "Token indices sequence length is longer than maximum (2108 > 2048)"

#### 8. **Flash Attention GPU Initialization** 
- **Problem**: "Flash Attention 2.0 with model not initialized on GPU" warning

### PR DESCRIPTION
**Pull request by Gayanukaa**
This PR adapts the VideoLLaMA3 repository for fine-tuning on the QVED (Physiotherapy Exercise Evaluation) dataset on a single NVIDIA GPU.

- Complete single-GPU training pipeline for VideoLLaMA3-2B
- Memory-optimized configuration (fits in 48GB VRAM)
- Automated inference and evaluation pipeline
- WandB integration for experiment tracking
- HuggingFace model upload support
- Training visualization and reporting tools

### 1. Dataset Format Adaptation

- Original dataset used string format for video paths: `"video": "path/to/video.mp4"`
- VideoLLaMA3 training expects list format: `"video": ["path/to/video.mp4"]`
- Missing `<video>` tag in conversation prompts

### 2. Single GPU Optimization

- Default configuration assumed 8 GPUs (NPROC_PER_NODE=8)
- Multi-GPU setup caused resource conflicts on single GPU
- Flash Attention initialization errors
- Out of memory errors with default parameters

**Key Changes:**

| Parameter             | Original   | Optimized      | Reason                                  |
| --------------------- | ---------- | -------------- | --------------------------------------- |
| NPROC_PER_NODE        | 8          | 1              | Single GPU setup                        |
| DeepSpeed Config      | ZeRO-3     | ZeRO-2         | Better for single GPU + Flash Attention |
| Batch Size            | 4 (global) | 1 (per-device) | Memory optimization                     |
| Gradient Accumulation | Variable   | 4              | Maintain effective batch size           |
| max_frames            | 180        | 32             | Reduce video memory usage               |
| model_max_length      | 16384      | 4096           | Reduce sequence memory                  |
| mm_max_length         | 10240      | 3072           | Reduce multimodal memory                |

### 3. Training Infrastructure

**A. [`scripts/finetune/finetune_qved.sh`](scripts/finetune/finetune_qved.sh)**

- Main fine-tuning script with all optimizations
- Automatic model download checking
- Dataset verification
- WandB logging integration
- Memory-optimized hyperparameters

**B. [`scripts/finetune/quickstart_finetune.sh`](scripts/finetune/quickstart_finetune.sh)**

- Complete training workflow automation
- Setup verification
- Training execution
- Automatic plot generation
- Optional HuggingFace upload

**C. [`scripts/finetune/verify_qved_setup.sh`](scripts/finetune/verify_qved_setup.sh)**

- Comprehensive pre-training checks
- Dataset integrity verification (supports list format video paths)
- Environment validation
- GPU availability check

### 4. Inference and Evaluation Pipeline

- Added VideoLLaMA3 parameters (fps, max_frames)
- Auto-detection of latest checkpoint
- Automatic evaluation report generation

#### A. [`utils/download_models.py`](utils/download_models.py)

- Pre-download VideoLLaMA3-2B and SigLIP-NaViT models
- Fallback to huggingface-cli if transformers download fails
- Progress reporting

#### B. Visualization and Reporting

**Existing Tools Integrated:**

- [`utils/plot_training_stats.py`](utils/plot_training_stats.py) - Training curve visualization
- [`utils/generate_test_report.py`](utils/generate_test_report.py) - Excel evaluation reports
- [`utils/hf_upload.py`](utils/hf_upload.py) - HuggingFace model upload

### 6. Logging and Monitoring

Removed TensorBoard → Added WandB

### Complete Training Workflow

```bash
# 1. Setup environment (first time only)
bash scripts/setup/setup_train.sh

# 2. Activate environment
conda activate mobile_videogpt

# 3. Initialize dataset
python utils/load_dataset.py
python utils/filter_ground_truth.py
python utils/qved_from_fined_labels.py

# 4. Run complete training workflow
bash scripts/finetune/quickstart_finetune.sh
# This will:
# - Verify setup
# - Download models (if needed)
# - Train model
# - Generate plots
# - Offer HF upload

# 5. Run inference on test set
bash scripts/inference/run_inference.sh --model_path results/qved_finetune
# Auto-detects latest checkpoint
# Generates predictions and evaluation report
```

### CODE REVIEW

**Review by Nidhan03 — COMMENTED** [review]
Everything looks good!
